// redux-async-utils/lib/actionCreators.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pendingActionCreator = pendingActionCreator;
exports.doneActionCreator = doneActionCreator;
exports.failureActionCreator = failureActionCreator;
exports.invalidateActionCreator = invalidateActionCreator;

var _constants = require('./constants');

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function pendingActionCreator(manageAsyncStateFor, group) {
  var _ref;

  return _ref = {
    type: _constants.PENDING
  }, _defineProperty(_ref, _constants.ASYNC_UTILS_MARKER, true), _defineProperty(_ref, _constants.ASYNC_UTILS_STATE_FOR, manageAsyncStateFor), _defineProperty(_ref, _constants.ASYNC_UTILS_STATE_GROUP, group), _defineProperty(_ref, _constants.ASYNC_UTILS_STATE, _constants.PENDING), _ref;
}

function doneActionCreator(manageAsyncStateFor) {
  var actionToDispatch = arguments.length <= 1 || arguments[1] === void 0 ? false : arguments[1];

  if (actionToDispatch) {
    actionToDispatch[_constants.ASYNC_UTILS_MARKER] = true;
    actionToDispatch[_constants.ASYNC_UTILS_STATE_FOR] = manageAsyncStateFor;
    actionToDispatch[_constants.ASYNC_UTILS_STATE] = _constants.DONE;
  } else {
    var _ref2;

    return _ref2 = {
      type: _constants.DONE
    }, _defineProperty(_ref2, _constants.ASYNC_UTILS_MARKER, true), _defineProperty(_ref2, _constants.ASYNC_UTILS_STATE_FOR, manageAsyncStateFor), _defineProperty(_ref2, _constants.ASYNC_UTILS_STATE, _constants.DONE), _ref2;
  }

  return actionToDispatch;
}

function failureActionCreator(manageAsyncStateFor) {
  var _ref3;

  var error = arguments.length <= 1 || arguments[1] === void 0 ? null : arguments[1];

  return _ref3 = {
    type: _constants.FAILURE,
    error: error
  }, _defineProperty(_ref3, _constants.ASYNC_UTILS_MARKER, true), _defineProperty(_ref3, _constants.ASYNC_UTILS_STATE_FOR, manageAsyncStateFor), _defineProperty(_ref3, _constants.ASYNC_UTILS_STATE, _constants.FAILURE), _ref3;
}

function invalidateActionCreator(manageAsyncStateFor) {
  var _ref4;

  return _ref4 = {
    type: _constants.INVALIDATED
  }, _defineProperty(_ref4, _constants.ASYNC_UTILS_MARKER, true), _defineProperty(_ref4, _constants.ASYNC_UTILS_STATE_FOR, manageAsyncStateFor), _defineProperty(_ref4, _constants.ASYNC_UTILS_STATE, _constants.INVALIDATED), _ref4;
}
