// redux-async-utils/lib/asyncActionsStateReducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.initialState = void 0;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = asyncActionsState;

var _constants = require('./constants');

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var initialState = exports.initialState = {
  asyncActionsStates: [],
  indexes: {},
  groups: {},
  errorsCount: 0,
  successCount: 0,
  failedActionsIndexes: [],
  digested: 0
};

function asyncActionsState() {
  var state = arguments.length <= 0 || arguments[0] === void 0 ? initialState : arguments[0];
  var action = arguments.length <= 1 || arguments[1] === void 0 ? {} : arguments[1];

  var newState = state;

  if (action[_constants.ASYNC_UTILS_MARKER]) {
    var _ret = function () {
      var asyncState = action[_constants.ASYNC_UTILS_STATE];
      var error = action.error || null;
      var asyncStateFor = action[_constants.ASYNC_UTILS_STATE_FOR];
      var asyncStateGroup = action[_constants.ASYNC_UTILS_STATE_GROUP];
      var previousState = void 0;

      var index = void 0;

      newState = {
        asyncActionsStates: state.asyncActionsStates.slice(0),
        indexes: _extends({}, state.indexes),
        groups: _extends({}, state.groups),
        errorsCount: state.errorsCount,
        failedActionsIndexes: state.failedActionsIndexes.slice(0),
        successCount: state.successCount,
        digested: state.digested
      };

      // look for asyncState in indexes
      if (typeof newState.indexes[asyncStateFor] !== 'undefined') {
        index = newState.indexes[asyncStateFor];
        previousState = _extends({}, newState.asyncActionsStates[index]);
      } else {
        // search in the array and create indexes
        newState.asyncActionsStates.forEach(function (v, i) {
          if (v[_constants.ASYNC_UTILS_STATE_FOR] === asyncStateFor) {
            index = i;
            previousState = _extends({}, newState.asyncActionsStates[index]);
          }

          newState.indexes[v[_constants.ASYNC_UTILS_STATE_FOR]] = i;
        });
      }

      // if a pending action become INVALIDATED before being completed,
      // when DONE or FAILURE state will be dispatched
      // simply ignore them
      if (previousState && asyncState !== _constants.PENDING && previousState.state === _constants.INVALIDATED) {
        return {
          v: newState
        };
      }

      // create new
      if (typeof index === 'undefined') {
        var _newActionState;

        var newActionState = (_newActionState = {}, _defineProperty(_newActionState, _constants.ASYNC_UTILS_STATE_FOR, asyncStateFor), _defineProperty(_newActionState, 'state', asyncState), _defineProperty(_newActionState, 'error', error), _newActionState);
        newState.asyncActionsStates.push(newActionState);

        // save index
        newState.indexes[asyncStateFor] = newState.asyncActionsStates.length - 1;
        index = newState.asyncActionsStates.length - 1;

        // save in group
        if (asyncStateGroup) {
          if (!newState.groups[asyncStateGroup]) {
            newState.groups[asyncStateGroup] = [index];
          } else {
            newState.groups[asyncStateGroup] = newState.groups[asyncStateGroup].slice(0);
            newState.groups[asyncStateGroup].push(index);
          }
        }
      } else {
        var _newState$asyncAction;

        newState.asyncActionsStates[index] = (_newState$asyncAction = {}, _defineProperty(_newState$asyncAction, _constants.ASYNC_UTILS_STATE_FOR, asyncStateFor), _defineProperty(_newState$asyncAction, 'state', asyncState), _defineProperty(_newState$asyncAction, 'error', error), _newState$asyncAction);
      }

      // if an async action was digested and a new pending action is requested,
      // reset digested, successCount and errorsCount as the action was never requested before
      if (previousState &&

      // prevent variables reset in case of simple passage from PENDING to other state
      // in that case the action has to be digested yet
      previousState.state !== _constants.PENDING &&

      // prevent pending action dispatched twice
      previousState.state !== asyncState) {
        if (asyncState !== _constants.INVALIDATED) {
          newState.digested--;
        }

        if (previousState.state === _constants.DONE) {
          newState.successCount--;
        }

        if (previousState.state === _constants.FAILURE) {
          newState.errorsCount--;
          newState.failedActionsIndexes.splice(newState.failedActionsIndexes.indexOf(index), 1);
        }
      }

      // update vars
      if (asyncState === _constants.INVALIDATED && (!previousState || previousState.state === _constants.PENDING)) {
        newState.digested++;
      }

      if (asyncState === _constants.DONE && (!previousState || previousState.state !== _constants.DONE)) {
        newState.digested++;
        newState.successCount++;
      }

      if (asyncState === _constants.FAILURE && (!previousState || previousState.state !== _constants.FAILURE)) {
        newState.digested++;
        newState.errorsCount++;
        newState.failedActionsIndexes.push(index);
      }
    }();

    if ((typeof _ret === 'undefined' ? 'undefined' : _typeof(_ret)) === "object") return _ret.v;
  }

  return newState;
}
